File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java
package com.airbnb.epoxy;

import android.graphics.Rect;
import android.view.View;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Px;
import androidx.recyclerview.widget.RecyclerView;

/**
 * This class represent an item in the {@link androidx.recyclerview.widget.RecyclerView} and it is
 * being reused with multiple model via the update method. There is 1:1 relationship between an
 * EpoxyVisibilityItem and a child within the {@link androidx.recyclerview.widget.RecyclerView}.
 *
 * It contains the logic to compute the visibility state of an item. It will also invoke the
 * visibility callbacks on {@link com.airbnb.epoxy.EpoxyViewHolder}
 *
 * This class should remain non-public and is intended to be used by {@link EpoxyVisibilityTracker}
 * only.
 */
class EpoxyVisibilityItem {

  private static final int NOT_NOTIFIED = -1;

  private final Rect localVisibleRect = new Rect();

  private int adapterPosition = RecyclerView.NO_POSITION;

  @Px
  private int height;
  @Px
  private int width;

  @Px
  private int visibleHeight;
  @Px
  private int visibleWidth;

  @Px
  private int viewportHeight;
  @Px
  private int viewportWidth;

  private boolean partiallyVisible = false;
  private boolean fullyVisible = false;
  private boolean visible = false;
  private boolean focusedVisible = false;

  /** Store last value for de-duping */
  private int lastVisibleHeightNotified = NOT_NOTIFIED;
  private int lastVisibleWidthNotified = NOT_NOTIFIED;

  EpoxyVisibilityItem(int adapterPosition) {
    reset(adapterPosition);
  }

  /**
   * Update the visibility item according the current layout.
   *
   * @param view        the current {@link com.airbnb.epoxy.EpoxyViewHolder}'s itemView
   * @param parent      the {@link androidx.recyclerview.widget.RecyclerView}
   * @return true if the view has been measured
   */
  boolean update(@NonNull View view, @NonNull RecyclerView parent, boolean detachEvent) {
    // Clear the rect before calling getLocalVisibleRect
    localVisibleRect.setEmpty();
    boolean viewDrawn = view.getLocalVisibleRect(localVisibleRect) && !detachEvent;
    height = view.getHeight();
    width = view.getWidth();
    viewportHeight = parent.getHeight();
    viewportWidth = parent.getWidth();
    visibleHeight = viewDrawn ? localVisibleRect.height() : 0;
    visibleWidth = viewDrawn ? localVisibleRect.width() : 0;
    return height > 0 && width > 0;
  }

  int getAdapterPosition() {
    return adapterPosition;
  }

  void reset(int newAdapterPosition) {
    fullyVisible = false;
    visible = false;
    focusedVisible = false;
    adapterPosition = newAdapterPosition;
    lastVisibleHeightNotified = NOT_NOTIFIED;
    lastVisibleWidthNotified = NOT_NOTIFIED;
  }

  void handleVisible(@NonNull EpoxyViewHolder epoxyHolder, boolean detachEvent) {
    boolean previousVisible = visible;
    visible = !detachEvent && isVisible();
    if (visible != previousVisible) {
      if (visible) {
        epoxyHolder.visibilityStateChanged(VisibilityState.VISIBLE);
      } else {
        epoxyHolder.visibilityStateChanged(VisibilityState.INVISIBLE);
      }
    }
  }

  void handleFocus(EpoxyViewHolder epoxyHolder, boolean detachEvent) {
    boolean previousFocusedVisible = focusedVisible;
    focusedVisible = !detachEvent && isInFocusVisible();
    if (focusedVisible != previousFocusedVisible) {
      if (focusedVisible) {
        epoxyHolder.visibilityStateChanged(VisibilityState.FOCUSED_VISIBLE);
      } else {
        epoxyHolder.visibilityStateChanged(VisibilityState.UNFOCUSED_VISIBLE);
      }
    }
  }

  void handlePartialImpressionVisible(EpoxyViewHolder epoxyHolder, boolean detachEvent,
      @IntRange(from = 0, to = 100) int thresholdPercentage) {
    boolean previousPartiallyVisible = partiallyVisible;
    partiallyVisible = !detachEvent && isPartiallyVisible(thresholdPercentage);
    if (partiallyVisible != previousPartiallyVisible) {
      if (partiallyVisible) {
        epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_VISIBLE);
      } else {
        epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_INVISIBLE);
      }
    }
  }

  void handleFullImpressionVisible(EpoxyViewHolder epoxyHolder, boolean detachEvent) {
    boolean previousFullyVisible = fullyVisible;
    fullyVisible = !detachEvent && isFullyVisible();
    if (fullyVisible != previousFullyVisible) {
      if (fullyVisible) {
        epoxyHolder.visibilityStateChanged(VisibilityState.FULL_IMPRESSION_VISIBLE);
      }
    }
  }

  boolean handleChanged(EpoxyViewHolder epoxyHolder, boolean visibilityChangedEnabled) {
    boolean changed = false;
    if (visibleHeight != lastVisibleHeightNotified || visibleWidth != lastVisibleWidthNotified) {
      if (visibilityChangedEnabled) {
        epoxyHolder.visibilityChanged(
            100.f / height * visibleHeight,
            100.f / width * visibleWidth,
            visibleHeight, visibleWidth
        );
      }
      lastVisibleHeightNotified = visibleHeight;
      lastVisibleWidthNotified = visibleWidth;
      changed = true;
    }
    return changed;
  }

  private boolean isVisible() {
    return visibleHeight > 0 && visibleWidth > 0;
  }

  private boolean isInFocusVisible() {
    final int halfViewportArea = viewportHeight * viewportWidth / 2;
    final int totalArea = height * width;
    final int visibleArea = visibleHeight * visibleWidth;
    // The model has entered the focused range either if it is larger than half of the viewport
    // and it occupies at least half of the viewport or if it is smaller than half of the viewport
    // and it is fully visible.
    return (totalArea >= halfViewportArea)
        ? (visibleArea >= halfViewportArea)
        : totalArea == visibleArea;
  }

  private boolean isPartiallyVisible(@IntRange(from = 0, to = 100) int thresholdPercentage) {
    if (thresholdPercentage == 0) return false;

    final int totalArea = height * width;
    final int visibleArea = visibleHeight * visibleWidth;
    final float visibleAreaPercentage = (visibleArea / (float) totalArea) * 100;

    return visibleAreaPercentage >= thresholdPercentage;
  }

  private boolean isFullyVisible() {
    return visibleHeight == height && visibleWidth == width;
  }

  void shiftBy(int offsetPosition) {
    adapterPosition += offsetPosition;
  }
}

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.kt
package com.airbnb.epoxy

import android.graphics.Rect
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IntRange
import androidx.annotation.Px
import androidx.annotation.VisibleForTesting
import androidx.recyclerview.widget.RecyclerView

/**
 * This class represent an item in a [android.view.ViewGroup] and it is
 * being reused with multiple model via the update method. There is 1:1 relationship between an
 * EpoxyVisibilityItem and a child within the [android.view.ViewGroup].
 *
 * It contains the logic to compute the visibility state of an item. It will also invoke the
 * visibility callbacks on [com.airbnb.epoxy.EpoxyViewHolder]
 *
 * This class should remain non-public and is intended to be used by [EpoxyVisibilityTracker]
 * only.
 */
@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
class EpoxyVisibilityItem(adapterPosition: Int? = null) {

    private val localVisibleRect = Rect()

    var adapterPosition = RecyclerView.NO_POSITION
        private set

    @Px
    private var height = 0

    @Px
    private var width = 0

    @Px
    private var visibleHeight = 0

    @Px
    private var visibleWidth = 0

    @Px
    private var viewportHeight = 0

    @Px
    private var viewportWidth = 0
    private var partiallyVisible = false
    private var fullyVisible = false
    private var visible = false
    private var focusedVisible = false
    private var viewVisibility = View.GONE

    /** Store last value for de-duping  */
    private var lastVisibleHeightNotified: Int? = null
    private var lastVisibleWidthNotified: Int? = null
    private var lastVisibilityNotified: Int? = null

    init {
        adapterPosition?.let {
            reset(it)
        }
    }

    /**
     * Update the visibility item according the current layout.
     *
     * @param view        the current [com.airbnb.epoxy.EpoxyViewHolder]'s itemView
     * @param parent      the [android.view.ViewGroup]
     * @return true if the view has been measured
     */
    fun update(view: View, parent: ViewGroup, detachEvent: Boolean): Boolean {
        // Clear the rect before calling getLocalVisibleRect
        localVisibleRect.setEmpty()
        val viewDrawn = view.getLocalVisibleRect(localVisibleRect) && !detachEvent
        height = view.height
        width = view.width
        viewportHeight = parent.height
        viewportWidth = parent.width
        visibleHeight = if (viewDrawn) localVisibleRect.height() else 0
        visibleWidth = if (viewDrawn) localVisibleRect.width() else 0
        viewVisibility = view.visibility
        return height > 0 && width > 0
    }

    fun reset(newAdapterPosition: Int) {
        fullyVisible = false
        visible = false
        focusedVisible = false
        adapterPosition = newAdapterPosition
        lastVisibleHeightNotified = null
        lastVisibleWidthNotified = null
        lastVisibilityNotified = null
    }

    fun handleVisible(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
        val previousVisible = visible
        visible = !detachEvent && isVisible()
        if (visible != previousVisible) {
            if (visible) {
                epoxyHolder.visibilityStateChanged(VisibilityState.VISIBLE)
            } else {
                epoxyHolder.visibilityStateChanged(VisibilityState.INVISIBLE)
            }
        }
    }

    fun handleFocus(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
        val previousFocusedVisible = focusedVisible
        focusedVisible = !detachEvent && isInFocusVisible()
        if (focusedVisible != previousFocusedVisible) {
            if (focusedVisible) {
                epoxyHolder.visibilityStateChanged(VisibilityState.FOCUSED_VISIBLE)
            } else {
                epoxyHolder.visibilityStateChanged(VisibilityState.UNFOCUSED_VISIBLE)
            }
        }
    }

    fun handlePartialImpressionVisible(
        epoxyHolder: EpoxyViewHolder,
        detachEvent: Boolean,
        @IntRange(from = 0, to = 100) thresholdPercentage: Int
    ) {
        val previousPartiallyVisible = partiallyVisible
        partiallyVisible = !detachEvent && isPartiallyVisible(thresholdPercentage)
        if (partiallyVisible != previousPartiallyVisible) {
            if (partiallyVisible) {
                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_VISIBLE)
            } else {
                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_INVISIBLE)
            }
        }
    }

    fun handleFullImpressionVisible(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
        val previousFullyVisible = fullyVisible
        fullyVisible = !detachEvent && isFullyVisible()
        if (fullyVisible != previousFullyVisible) {
            if (fullyVisible) {
                epoxyHolder.visibilityStateChanged(VisibilityState.FULL_IMPRESSION_VISIBLE)
            }
        }
    }

    fun handleChanged(epoxyHolder: EpoxyViewHolder, visibilityChangedEnabled: Boolean): Boolean {
        var changed = false
        if (visibleHeight != lastVisibleHeightNotified || visibleWidth != lastVisibleWidthNotified || viewVisibility != lastVisibilityNotified) {
            if (visibilityChangedEnabled) {
                if (viewVisibility == View.GONE) {
                    epoxyHolder.visibilityChanged(0f, 0f, 0, 0)
                } else {
                    epoxyHolder.visibilityChanged(
                        100f / height * visibleHeight,
                        100f / width * visibleWidth,
                        visibleHeight, visibleWidth
                    )
                }
            }
            lastVisibleHeightNotified = visibleHeight
            lastVisibleWidthNotified = visibleWidth
            lastVisibilityNotified = viewVisibility
            changed = true
        }
        return changed
    }

    private fun isVisible(): Boolean {
        return viewVisibility == View.VISIBLE && visibleHeight > 0 && visibleWidth > 0
    }

    private fun isInFocusVisible(): Boolean {
        val halfViewportArea = viewportHeight * viewportWidth / 2
        val totalArea = height * width
        val visibleArea = visibleHeight * visibleWidth
        // The model has entered the focused range either if it is larger than half of the viewport
        // and it occupies at least half of the viewport or if it is smaller than half of the viewport
        // and it is fully visible.
        return viewVisibility == View.VISIBLE &&
            if (totalArea >= halfViewportArea) visibleArea >= halfViewportArea else totalArea == visibleArea
    }

    private fun isPartiallyVisible(
        @IntRange(
            from = 0,
            to = 100
        ) thresholdPercentage: Int
    ): Boolean {
        // special case 0%: trigger as soon as some pixels are one the screen
        if (thresholdPercentage == 0) return isVisible()
        val totalArea = height * width
        val visibleArea = visibleHeight * visibleWidth
        val visibleAreaPercentage = visibleArea / totalArea.toFloat() * 100
        return viewVisibility == View.VISIBLE && visibleAreaPercentage >= thresholdPercentage
    }

    private fun isFullyVisible(): Boolean {
        return viewVisibility == View.VISIBLE && visibleHeight == height && visibleWidth == width
    }

    fun shiftBy(offsetPosition: Int) {
        adapterPosition += offsetPosition
    }
}
